fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -51,20 +54,79 @@
+
+            if (compartment != null) {
+                // Compartment search - validate patient access
+                if (!"Patient".equals(compartment)) {
+                    String msg = "Compartment search with compartment '" + compartment + "' is not permitted.";
+                    throw new FHIRPersistenceInterceptorException(msg).withIssue(FHIRUtil.buildOperationOutcomeIssue(msg, IssueType.FORBIDDEN));
+                }
+                if (!patientIdFromToken.contains(compartmentId)) {
+                    String msg = "Interaction with 'Patient/" + compartmentId + "' is not permitted under patient context '" + patientIdFromToken + "'.";
+                    throw new FHIRPersistenceInterceptorException(msg).withIssue(FHIRUtil.buildOperationOutcomeIssue(msg, IssueType.FORBIDDEN));
+                }
+                try {
+                    if (!CompartmentUtil.getCompartmentResourceTypes("Patient").contains(event.getFhirResourceType())) {
+                        String msg = "Resource type '" + event.getFhirResourceType() + "' is not valid for Patient compartment search.";
+                        throw new FHIRPersistenceInterceptorException(msg).withIssue(FHIRUtil.buildOperationOutcomeIssue(msg, IssueType.INVALID));
+                    }
+                } catch (FHIRSearchException e) {
+                    log.log(Level.WARNING, "Unexpected exception while getting compartment resource types", e);
+                }
+            } else {
+                // Not compartment search - validate and convert to Patient compartment search if resource type is member of Patient compartment
+                if ("Patient".equals(event.getFhirResourceType())) {
+                    String msg = "Non-compartment Patient search is not permitted.";
+                    throw new FHIRPersistenceInterceptorException(msg).withIssue(FHIRUtil.buildOperationOutcomeIssue(msg, IssueType.FORBIDDEN));
+                }
+                try {
+                    if (CompartmentUtil.getCompartmentResourceTypes("Patient").contains(event.getFhirResourceType())) {
+                        // NOTE: We currently do not support OR'd compartment searches, nor do we currently expect more than one patient ID to be
+                        // specified in the authorization token (see getPatientIdFromToken()). We will use the first ID specified for the compartment search.
+                        FHIRSearchContext compartmentSearchContext = SearchUtil.parseQueryParameters("Patient", patientIdFromToken.get(0),
+                            ModelSupport.getResourceType(event.getFhirResourceType()), Collections.emptyMap(), searchContext.isLenient());
+                        searchContext.getSearchParameters().addAll(compartmentSearchContext.getSearchParameters());
+                    }
+                } catch (Exception e) {
+                    log.log(Level.WARNING, "Unexpected exception while converting search to Patient compartment search", e);
```

Comment:
I'm thinking we should wrap and rethrow in this case.  If not, then we would continue to execute the search and could potentially leak whether the resources exist or not.  Granted, we should never get here anyway, so its not too critical.

Reply:
Done - wrap and rethrow.